working/macros/api/code.dart

```diff
@@ -0,0 +1,129 @@
+/// A scope in which to resolve a chunk of code.
+///
+/// TODO: Handle more deeply nested scopes (such as a scope specific to a class
+/// or method body).
+class Scope {
+  /// Identifiers should be resolved as if they existed in this library.
+  final Uri libraryUri;
+
+  Scope(this.libraryUri);
```

Comment:
I don't think users should be able to create a Scope for an arbitrary URI. Instead, they should only be able to get access to a couple of existing Scope instances through the macro API: The top-level Scope where the macro is declared and maybe one where the macro is applied.

Reply:
Ya, I think ultimately I do agree with you. I wanted to expose this so that macros could add references to things they don't directly import - but on second that thought would create weird situations with dependencies, especially in google3 but even in pub as well. It could in theory add a reference to something that isn't depended on at all, forcing users to explicitly add their own dependency on something they don't use.